pkg/kv/transport.go

```diff
@@ -215,19 +214,26 @@ func (gt *grpcTransport) MoveToFront(replica roachpb.ReplicaDescriptor) {
 	defer gt.clientPendingMu.Unlock()
 	for i := range gt.orderedClients {
 		if gt.orderedClients[i].args.Replica == replica {
-			// If a call to this replica is active or retried, don't move it.
-			if gt.orderedClients[i].pending || gt.orderedClients[i].retried {
+			// If a call to this replica is active, don't move it.
+			if gt.orderedClients[i].pending {
 				return
 			}
 			// If we've already processed the replica, decrement the current
-			// index before we swap.
+			// index and swap.
 			if i < gt.clientIndex {
-				gt.orderedClients[i].retried = true
 				gt.clientIndex--
+				gt.orderedClients[i], gt.orderedClients[gt.clientIndex] =
+					gt.orderedClients[gt.clientIndex], gt.orderedClients[i]
+			} else {
+				// Otherwise, move the replica to the front and translate
+				// successive replicas to the rear.
+				newOrderedClients := make([]batchClient, 0, len(gt.orderedClients))
+				newOrderedClients = append(newOrderedClients, gt.orderedClients[:gt.clientIndex]...)
+				newOrderedClients = append(newOrderedClients, gt.orderedClients[i])
+				newOrderedClients = append(newOrderedClients, gt.orderedClients[gt.clientIndex:i]...)
+				newOrderedClients = append(newOrderedClients, gt.orderedClients[i+1:]...)
+				gt.orderedClients = newOrderedClients
```

Comment:
This is an unusual code path, so I'm not worried about allocations. I prefer the explicit and verbose code because as you're discovering, avoiding the allocation is tricky.